src/host/telemetry/telemetry.hpp

```diff
@@ -0,0 +1,99 @@
+/*
+ *    Copyright (c) 2025, The OpenThread Authors.
+ *    All rights reserved.
+ *
+ *    Redistribution and use in source and binary forms, with or without
+ *    modification, are permitted provided that the following conditions are met:
+ *    1. Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *    2. Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *    3. Neither the name of the copyright holder nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ *    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *    ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef OTBR_TELEMETRY_HPP_
+#define OTBR_TELEMETRY_HPP_
+
+#include "openthread-br/config.h"
+
+#if OTBR_ENABLE_TELEMETRY_DATA_API
+
+#include <openthread/instance.h>
+
+#include "mdns/mdns.hpp"
+#include "proto/thread_telemetry.pb.h"
+
+#include "telemetry_retriever_border_agent.hpp"
+
+namespace otbr {
+namespace Host {
+
+class TelemetryRetriever
+{
+public:
+    /**
+     * Constructor.
+     *
+     * @param[in] aInstance  A pointer to the OT instance.
+     */
+    explicit TelemetryRetriever(otInstance *aInstance);
```

Comment:
Since you're passing an OT instance into it, is this specific to RCP?

Reply:
Yes, currently it is.

Reply:
What's the plan of expanding it to cover NCP?

Reply:
No plan yet. I considered it before. But the implementation in NCP mode will be very different. (We can not use OT APIs here to fetch the telemetry) We will probably use a different class to fetch the telemetry from NCP.